### PR TITLE
Add Differentiable Particle Filter Support

### PR DESCRIPTION
**Pull request by DanWaxman**
This PR adds differentiable particle filter support. These entail a general `ContDiscreteNonlinearSSM` class, which allows for non-Gaussian, non-linear observation processes. We include a differentiable particle filter using either "soft" or "stop gradient" resampling methods. 

These methods are integrated into the Lorenz-63 example notebook (the new API one), as well as a new `dpf_examples.ipynb` notebook, which shows a few examples of non-Gaussian observations (an OU-driven count process with Poisson observations, and a tracking example with bearing (S^1) and signal power (R) observations.

Remaining to-dos:
- [x] Add more documentation, including some citations to the literature
- [x] Clean up code after documentation

Two things that could be done in a follow-up PR, IMO:
- Include discrete-discrete processes
- Add in other resampling strategies, and potentially resampling criterion (c.f. [resampling via OT](https://proceedings.mlr.press/v139/corenflos21a.html), or [other ESS approximaitons](https://link.springer.com/article/10.1007/s00180-025-01665-8)).

### CODE REVIEW

File: cd_dynamax/src/continuous_discrete_nonlinear_ssm/cdnlssm_utils.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+        base_sample = self.base_distribution.sample(x, u, t)
+        return self.transform.f(base_sample, u, t)
+
+
+# Dynamics container for CD-NLSSM
+class ParamsCDNLSSMDynamics(NamedTuple):
+    drift: LearnableFunction
+    diffusion_coefficient: LearnableFunction
+    diffusion_cov: LearnableFunction
+    approx_order: Union[float, ParameterProperties]
```
Comment (DanWaxman) [inline cd_dynamax/src/continuous_discrete_nonlinear_ssm/cdnlssm_utils.py]: Use `ParamsCDNLGSSMDynamics` instead?

File: cd_dynamax/src/continuous_discrete_nonlinear_ssm/cdnlssm_utils.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    r"""Parameters of a nonlinear Gaussian SSM.
+
+    :param initial: initial distribution parameters
+    :param dynamics: dynamics distribution parameters
+    :param emissions: emission distribution parameters
+
+    The assumed transition and emission distributions are
+    $$p(z_1) = N(z_1 | m, S)$$
+    $$p(z_t | z_{t-1}, u_t) = N(z_t | m_t, P_t)$$
+    $$p(y_t | z_t) = N(y_t | h(z_t, u_t), R_t)$$
```
Comment (DanWaxman) [inline cd_dynamax/src/continuous_discrete_nonlinear_ssm/cdnlssm_utils.py]: Update these!

File: cd_dynamax/src/continuous_discrete_nonlinear_ssm/inference_dpf.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+def _multinomial_resample(key, x, log_w):
+    probs = jnp.exp(log_w - logsumexp(log_w))
+    idx = jr.choice(key, x.shape[0], shape=(x.shape[0],), p=probs, replace=True)
+    x_resampled = x[idx]
+    log_w_resampled = jnp.full_like(log_w, -jnp.log(x.shape[0]))
+    return x_resampled, log_w_resampled, idx
+
+
+def _stop_gradient_resample(key, x, log_w, *, base_method: str = "multinomial"):
+    """Resample while passing gradients through the chosen particles."""
```
Comment (DanWaxman) [inline cd_dynamax/src/continuous_discrete_nonlinear_ssm/inference_dpf.py]: Add references here!

File: cd_dynamax/src/continuous_discrete_nonlinear_ssm/inference_dpf.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    soft_weight = jnp.logaddexp(log_w + log_softness, log_uniform_component)
+    probs = jnp.exp(soft_weight - logsumexp(soft_weight))
+    idx = jr.choice(key, n, shape=(n,), p=probs, replace=True)
+    x_resampled = x[idx]
+    log_w_resampled = log_w[idx] - soft_weight[idx] - log_n
+    log_w_resampled = log_w_resampled - logsumexp(log_w_resampled)
+    return x_resampled, log_w_resampled
+
+
+def filter_dpf(
```
Comment (DanWaxman) [inline cd_dynamax/src/continuous_discrete_nonlinear_ssm/inference_dpf.py:225]: Check this is the same naming convention as EnKF etc.

File: cd_dynamax/src/continuous_discrete_nonlinear_ssm/inference_dpf.py
Comment (DanWaxman) [inline cd_dynamax/src/continuous_discrete_nonlinear_ssm/inference_dpf.py:1]: More explicit that this is a bootstrap PF

File: cd_dynamax/src/continuous_discrete_nonlinear_ssm/models.py [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+def compute_pushforward(
+    x0: Array,
+    P0: Array,
+    params: ParamsCDNLSSM,
+    t0: Float,
+    t1: Float,
+    inputs: Optional[Array] = None,
+    diffeqsolve_settings: Optional[dict] = None,
+) -> Tuple[Array, Array]:
+    """Propagate mean/covariance using chosen approximation order."""
```
Comment (DanWaxman) [inline cd_dynamax/src/continuous_discrete_nonlinear_ssm/models.py]: Dispatch to CDNLGSSM case if using Brownian motions, give errors otherwise.

Comment (DanWaxman) [reply cd_dynamax/src/continuous_discrete_nonlinear_ssm/models.py]: We don't really have a nice setup for this right now. I would propose a refactor that takes a lot of common utilities (like `compute_pushforward`) to `SDEs.py`, or something similar, though this seems out-of-scope for the current PR.

**Review by mattlevine22 — APPROVED** [review]
Looks great, thanks Dan!
